src/backend/distributed/operations/worker_shard_copy.c

```diff
@@ -344,21 +345,99 @@ ShardCopyDestReceiverDestroy(DestReceiver *dest)
 }
 
 
+static StringInfo
+ConstructTargetColumnsList(TupleDesc tupdesc)
+{
+	StringInfo columnList = makeStringInfo();
+	int i;
+
+	for (i = 0; i < tupdesc->natts; i++)
+	{
+		Form_pg_attribute att = TupleDescAttr(tupdesc, i);
+		if (att->attgenerated || att->attisdropped)
+		{
+			continue;
+		}
+		if (i > 0)
+		{
+			appendStringInfo(columnList, ", ");
+		}
+
+		appendStringInfo(columnList, "%s ", NameStr(att->attname));
+	}
+
+	elog(WARNING, "ConstructTargetColumnsList : %s", columnList->data);
+
+	return columnList;
+}
+
+
+StringInfo
+ConstructNonGeneratedColumnsList(const char *shardRelationName, const char *schemaName)
+{
+	elog(WARNING, "ConstructNonGeneratedColumnsList : %s, %s", shardRelationName,
+		 schemaName);
+
+	Oid namespaceOid = get_namespace_oid(schemaName, true);
+
+	Oid relationId = get_relname_relid(shardRelationName, namespaceOid);
+
+	Relation relation = relation_open(relationId, AccessShareLock);
+	int numberOfAttributes = RelationGetNumberOfAttributes(relation);
+
+	StringInfo columnList = makeStringInfo();
+
+	bool first = true;
+
+	for (int attrNum = 1; attrNum <= numberOfAttributes; attrNum++)
+	{
+		Form_pg_attribute attributeTuple =
+			TupleDescAttr(relation->rd_att, attrNum - 1);
+
+		if (attributeTuple->attgenerated || attributeTuple->attisdropped)
+		{
+			continue;
+		}
+
+		if (!first)
+		{
+			appendStringInfo(columnList, ", ");
+		}
+
+		appendStringInfo(columnList, "%s ", NameStr(attributeTuple->attname));
+
+		first = false;
+	}
+
```

Comment:
Instead of implementing the same loop twice, I think this can call `ConstructTargetColumnsList(relation->rd_att)`